I fixed it! no more HasSignature hanging (+ performance improv idea)

LucasParsy opened the pull request:
Hi!

After so much time on this issue, I finally solved it, you just had to call HasSignature from outside of the `OnDLLNotification` callback, so now the callback only fills a list that the main `Detections::Monitor` loop checks. It may delay the check by a few seconds, but it won't hang anymore 😃 


#### performance improvement ideas

Also, I wanted to understand why the Anti-Cheat constructor took 80 seconds to load for me, and see if we could *"optimize"* (understand : put in threads) some parts. I think making the main thread hanging a minute is not a desired behavior :/

And good news, Visual Studio generated me a 20Go report file to tell me only 2 lines of code take 98% of the start time:


![profilerAntiCheat](https://github.com/user-attachments/assets/c0fae2dc-313f-471b-862c-850cda3b180d)

88%:
https://github.com/AlSch092/UltimateAntiCheat/blob/b095897a57c8c685822186057593e6dc9c83c99b/AntiTamper/Integrity.hpp#L47

10%:
https://github.com/AlSch092/UltimateAntiCheat/blob/b095897a57c8c685822186057593e6dc9c83c99b/API/API.cpp#L147

The idea is that these two methods populate variables in their constructor, why not put them in a thread and just add a "isDataReady" boolean flag so that methods needing them can wait/do something else ? this way, even no need of mutexes!

It seems simple to implement (I can do it, but I want your input on the subject):
for `GetUnsignedDrivers`, it's even unused for now, so we can decide what to do.
for `GetModuleHashes`, the content (`Integrity::ModuleHashes`) is only used once, :

https://github.com/AlSch092/UltimateAntiCheat/blob/b095897a57c8c685822186057593e6dc9c83c99b/AntiTamper/Integrity.cpp#L189


https://github.com/AlSch092/UltimateAntiCheat/blob/b095897a57c8c685822186057593e6dc9c83c99b/Detections.cpp#L194

so as it's a monitor main loop check, we could make `IsModuleModified` return "false" while its module list is not initialized, so it will delay only by a few seconds the tamper detection? (and no need for mutexes)

And for the destructors, we could add a boolean to break the `GetModuleHashes` loop in case of premature destruction, and win lot of time when debugging.

(or screw that, we wait for the data to be ready before destructing it, what we want to improve is *boot* time, not closing time, we can debug by Ctrl-C the app and users won't launch their game for less than a minute anyway. ^^' ) 

what do you think of this idea?

also minor nitpick, why is this list a pointer instead of a classic variable?
https://github.com/AlSch092/UltimateAntiCheat/blob/b095897a57c8c685822186057593e6dc9c83c99b/AntiTamper/Integrity.hpp#L107

LucasParsy committed:
Fixed the HasSignature hanging for OnDllNotification

move `HasSignature` call from the callback to the main detection loop.

no more hanging!

(with a shared queue and a mutex to be sure, even if I don't know if callbacks need mutexes check)

LucasParsy committed:
change cin use in main loop

cin awaits for a character then to press enter, so you had to press "q" *then* Enter, it was a bit confusing, so using `_kbhit` to quit on any key press

AlSch092 commented:
Thanks for putting the extra work in to try and fix the hang issue, I'll try to test it on a w11 machine soon.

As for taking a very long time in GetModuleHashes, this may be because it's generating a large list of uint64_t hashes for every single module loaded. Each vector/list has about 80k entries, so 80k * nModules is quite a large amount. The hashing process can be changed to a single uint64_t instead of using vectors, but it will no longer be possible to determine exactly where a module was modified (the index of the vector could be used currently to determine roughly the mem address which was edited by a cheater). The "GetStackedHashes" routine should be already in the project somewhere, and can make a hash using a single uint64 by a simple sum of all bytes.

As for unsigned drivers, this is quite important because in competitive games, drivers are the main method of cheating. The majority of people who have started cheating within the past few years all use drivers (especially for shooter games) and almost never usermode programs for cheats. Almost all paid cheats will also be drivers or bootkits. Since it's rare that these people will have EV certs to sign their driver, checking unsigned drivers is needed for spotting driver-based cheats which have been loaded. In a typical windows install, there are almost no unsigned drivers usually loaded. The slow execution may be related to cert checking each loaded driver, since usually a web request is made each time a cert is checked (if it's not cached already).

When I'm at the computer I'll look over this further

Cheers

AlSch092 commented:
Thanks for putting the extra work in to try and fix the hang issue, I'll try to test it on a w11 machine soon.

As for taking a very long time in GetModuleHashes, this may be because it's generating a large list of uint64_t hashes for every single module loaded. Each vector/list has about 80k entries, so 80k * nModules is quite a large amount. The hashing process can be changed to a single uint64_t instead of using vectors, but it will no longer be possible to determine exactly where a module was modified (the index of the vector could be used currently to determine roughly the mem address which was edited by a cheater).

As for unsigned modules, this is quite important because in competitive games, drivers are the main method of cheating. The majority of people who have started cheating within the past few years all use drivers (especially for shooter games) and almost never usermode programs for cheats. Almost all paid cheats will also be drivers or bootkits. Since it's rare that these people will have EV certs to sign their driver, checking unsigned drivers is needed for spotting driver-based cheats which have been loaded. In a typical windows install, there are almost no unsigned drivers usually loaded. The slow execution may be related to cert checking each loaded driver, since usually a web request is made each time a cert is checked (if it's not cached already).

When I'm at the computer I'll look over this further

Cheers

AlSch092 reviewed (commented):
Looks mostly fine, however a couple of things:

1. Instead of a queue<LPCWSTR> , we should use queue<wstring> and push a copy of the LPCWSTR as a wstring into the queue to prevent possible memory corruption/dangling pointers. Since LPCWSTR is a pointer to a wchar_t*, it's possible it can become unallocated before we can process it, which would result in an unhandled exception.

2. using lock & unlock with the mutex can be replaced with a lock_guard, and the routine checkDLLSignature() does too much work while locked, which can result in the OnDllNotification routine being blocked if ::HasSignature takes too long to execute (due to possibly a network slowdown or any other reason). The lock should only be used while obtaining the queue entry, not until all queue items have been processed. If ::HasSignature ever hangs then it would create an error where OnDllNotification can no longer return since its waiting on a lock to be unlocked (which won't be unlocked if ::HasSignature takes too long to execute).

3. As for using _kbhit() in main, I think it's better we stick with std::cin because _kbhit() is not supported in certain compilers such as CLang/LLVM and will result in needing to make changes back to std::cin for other branches of the code (such as the second branch, which uses CLang)

AlSch092 committed:
Update main.cpp

There was a "return 0" before your main loop, which would have ended the thread immediately. we should stick to using std::cin for cross-compiler compatibility

AlSch092 committed:
change queue to wstring, remove std::

AlSch092 committed:
Update Detections.cpp

AlSch092 committed:
Update Detections.cpp

add nullptr check on monitor

LucasParsy commented:
Hi, thanks again for your input.

I should have noticed your point 1. and I was certain I did 2.  , but I did not. For 3 I genuinly did not know.
So sorry for the subpar code this time, I guess I was tired, I'll double check my pull reviews next time!

for the GetModuleHashes/GetUnsignedDrivers , interesting explanation, but I was just asking if offloading the main thread by putting these method calls in their own thread would be a good idea ^^'

Again sorry  but my ideas tend more on usability than new functionalities, and I find having the `AntiCheat` constructor taking a minute is not very practical to developers, and they would put it's initialization in a separate thread anyway, so why not do it ourself?

LucasParsy commented:
Ah, and now that I am seriously thginking about creating a Godot module... It looks more complicated than expected ^^'

https://docs.godotengine.org/en/stable/about/faq.html#why-does-godot-not-use-stl-standard-template-library

https://docs.godotengine.org/en/stable/about/faq.html#why-does-godot-not-use-exceptions

AlSch092 commented:
> Hi, thanks again for your input.
> 
> I should have noticed your point 1. and I was certain I did 2.  , but I did not. For 3 I genuinly did not know.
> So sorry for the subpar code this time, I guess I was tired, I'll double check my pull reviews next time!
> 
> for the GetModuleHashes/GetUnsignedDrivers , interesting explanation, but I was just asking if offloading the main thread by putting these method calls in their own thread would be a good idea ^^'
> 
> Again sorry  but my ideas tend more on usability than new functionalities, and I find having the `AntiCheat` constructor taking a minute is not very practical to developers, and they would put it's initialization in a separate thread anyway, so why not do it ourself?

No problem, it's not super obvious to do those points without having c++ as your main language. I'm certainly not perfect either at it, I started off the project using plain C for most of my coding years so getting used to C++ has been a journey here also.

I got too tired yesterday and didn't have the effort anymore to implement any fixes for the slow load times, I'll see if I can do this today. But probably yes we can offload any slow work into its own thread, it's better that way overall. It's odd that it takes so long to load for you, since I would assume you have a better processor than me if yours can support windows 11, mine is a intel 6700k and loads the program in about 5 seconds.

AlSch092 commented:
Making the project into Godot probably won't be easy, you would likely have to re-code all the concepts instead of just changing syntax. If you're able to call WINAPIs from Godot then it should be possible to make something similar, since the project relies heavily on winapis. You can probably implement higher-level concepts like integrity checking, but perhaps any low-level tricks won't be possible. Were you planning on making it a plug-in for Godot? (Never used it here)
